indexer/SymbolFormatter.cc

```diff
@@ -40,44 +41,197 @@ std::string_view SymbolFormatter::getMacroSymbol(clang::SourceLocation defLoc) {
+  auto symbol = getSymbol();
+  if (symbol.empty()) {
+    return "";
   }
-  std::string out{
-      fmt::format("c . todo-pkg todo-version {}/", this->scratchBuffer)};
-  this->scratchBuffer.clear();
   auto [newIt, inserted] =
-      this->declBasedCache.insert({namespaceDecl, std::move(out)});
+      this->declBasedCache.insert({decl, std::move(symbol)});
   ENFORCE(inserted);
   return std::string_view(newIt->second);
 }
 
+std::optional<std::string_view>
+SymbolFormatter::getContextSymbol(const clang::DeclContext *declContext) {
+  if (auto namespaceDecl = llvm::dyn_cast<clang::NamespaceDecl>(declContext)) {
+    return this->getNamespaceSymbol(namespaceDecl).value_or("");
+  }
+  if (auto tagDecl = llvm::dyn_cast<clang::TagDecl>(declContext)) {
+    return this->getTagSymbol(tagDecl);
+  }
+  if (llvm::isa<clang::TranslationUnitDecl>(declContext)
+      || llvm::isa<clang::ExternCContextDecl>(declContext)) {
+    auto decl = llvm::dyn_cast<clang::Decl>(declContext);
+    return this->getSymbolCached(
+        decl, [&]() -> std::string { return "c . todo-pkg todo-version "; });
+  }
+  // TODO: Handle all cases of DeclContext here:
+  // Done
+  // - TranslationUnitDecl
+  // - ExternCContext
+  // - NamespaceDecl
+  // - TagDecl
+  // Pending:
+  // - OMPDeclareReductionDecl
+  // - OMPDeclareMapperDecl
+  // - FunctionDecl
+  // - ObjCMethodDecl
+  // - ObjCContainerDecl
+  // - LinkageSpecDecl
+  // - ExportDecl
+  // - BlockDecl
+  // - CapturedDecl
+  return std::nullopt;
+}
+
+std::optional<std::string_view>
+SymbolFormatter::getTagSymbol(const clang::TagDecl *tagDecl) {
+  return this->getSymbolCached(tagDecl, [&]() -> std::string {
+    auto optContextSymbol = this->getContextSymbol(tagDecl->getDeclContext());
+    if (!optContextSymbol.has_value()) {
+      return {};
+    }
+    auto contextSymbol = optContextSymbol.value();
+    if (!tagDecl->getDeclName().isEmpty()) {
+      llvm::raw_string_ostream os(this->scratchBuffer);
+      static_cast<const clang::NamedDecl *>(tagDecl)->printName(os);
+      std::string out{fmt::format("{}{}#", contextSymbol, this->scratchBuffer)};
+      this->scratchBuffer.clear();
+      return out;
+    }
+    auto definitionTagDecl = tagDecl->getDefinition();
+    ENFORCE(definitionTagDecl, "can't forward-declare an anonymous type");
+    auto defLoc =
+        this->sourceManager.getExpansionLoc(definitionTagDecl->getLocation());
+
+    auto defFileId = this->sourceManager.getFileID(defLoc);
+    ENFORCE(defFileId.isValid());
+    auto counter = this->anonymousTypeCounters[{defFileId}]++;
+
+    auto declContext = definitionTagDecl->getDeclContext();
+    if (llvm::isa<clang::NamespaceDecl>(declContext)
+        || llvm::isa<clang::TranslationUnitDecl>(declContext)) {
+      // If the anonymous type is inside a namespace, then we know the
+      // DeclContext chain only has namespaces (types cannot contain
+      // namespaces). So include the file hash too, to avoid collisions across
```

Comment:
file*name* hash

Reply:
Fixed in [`0635dc4` (#104)](https://github.com/sourcegraph/scip-clang/pull/104/commits/0635dc4f47b7558bb6a9b75e3e9863ae465ad3a0)